src/common/stack_mmap.c

```diff
@@ -0,0 +1,342 @@
+		free_stack(mmap_stack_desc);
+	/* free local attr if used */
+	if (new_attr != ABT_THREAD_ATTR_NULL)
+		ABT_thread_attr_free(&new_attr);
+	return rc;
+}
+
+int mmap_stack_thread_create_on_xstream(ABT_xstream xstream,
+					void (*thread_func)(void *),
+					void *thread_arg, ABT_thread_attr attr,
+					ABT_thread *newthread)
+{
+	ABT_thread_attr new_attr = ABT_THREAD_ATTR_NULL;
+	int rc;
+	void *stack;
+	mmap_stack_desc_t *mmap_stack_desc = NULL;
+	size_t stack_size = MMAPED_ULT_STACK_SIZE, new_stack_size;
+	uint64_t shadow_alloced_stacks, shadow_free_stacks;
+
+	if (attr != ABT_THREAD_ATTR_NULL) {
+		ABT_thread_attr_get_stack(attr, &stack, &stack_size);
+		if (stack != NULL) {
+			/* an other external stack allocation method is being
+			 *  used, nothing to do
+			 */
+			rc = ABT_thread_create_on_xstream(xstream, thread_func,
+							  thread_arg, attr,
+							  newthread);
+			return rc;
+		}
+		if (stack_size < MMAPED_ULT_STACK_SIZE)
+			stack_size = MMAPED_ULT_STACK_SIZE;
+	} else {
+		rc = ABT_thread_attr_create(&new_attr);
+		if (rc != ABT_SUCCESS) {
+			D_ERROR("Create ABT thread attr failed: %d\n", rc);
+			return rc;
+		}
+	}
+
+	D_MUTEX_LOCK(&stack_free_list_lock);
+	shadow_alloced_stacks = alloced_stacks;
+	shadow_free_stacks = free_stacks;
+	if (!d_list_empty(&stack_free_list)) {
+		d_list_t *cur_stack;
+
+		D_ASSERT(free_stacks != 0);
+		d_list_for_each(cur_stack, &stack_free_list) {
+			mmap_stack_desc = container_of(cur_stack,
+						       mmap_stack_desc_t,
+						       stack_list);
+			if (mmap_stack_desc->stack_size >= stack_size)
+				break;
+		}
+		if (cur_stack != &stack_free_list) {
+			d_list_del_init(cur_stack);
+		} else {
+			D_MUTEX_UNLOCK(&stack_free_list_lock);
+			D_DEBUG(DB_MEM, "no stack of size >= %zd found on free list\n",
+				stack_size);
+			goto mmap_alloc;
+		}
+
+		--free_stacks;
+		D_MUTEX_UNLOCK(&stack_free_list_lock);
+		stack = mmap_stack_desc->stack;
+		stack_size = mmap_stack_desc->stack_size;
+		D_DEBUG(DB_MEM, "%p mmap()'ed stack of size %zd has been taken from free list, with alloced_stacks="DF_U64" and free_stacks="DF_U64"\n",
+			stack, stack_size, shadow_alloced_stacks,
+			shadow_free_stacks);
+	} else {
+		D_ASSERT(free_stacks == 0);
+mmap_alloc:
+		++alloced_stacks;
+		D_MUTEX_UNLOCK(&stack_free_list_lock);
+		stack = mmap(NULL, stack_size, PROT_READ | PROT_WRITE,
+			     MAP_PRIVATE | MAP_ANONYMOUS | MAP_STACK | MAP_GROWSDOWN,
+			     -1, 0);
+		if (stack == MAP_FAILED) {
+			D_ERROR("Failed to mmap() ULT stack of size %zd : %s, with alloced_stacks="DF_U64" and free_stacks="DF_U64"\n",
+				stack_size, strerror(errno),
+				shadow_alloced_stacks, shadow_free_stacks);
+			/* should be safe to decrease with no lock here */
+			--alloced_stacks;
+			/* return an ABT error */
+			D_GOTO(out_err, rc = ABT_ERR_MEM);
+		}
+
+		/* put descriptor at bottom of mmap()'ed stack */
+		mmap_stack_desc = (mmap_stack_desc_t *)(stack + stack_size -
+				  sizeof(mmap_stack_desc_t));
```

Comment:
I'm confused by this overall (MAP_GROWSDOWN is new to me) and I think some additional comments on what is going on here would help.

I don't fully understand what mmap() is returning here, even after reading https://man7.org/linux/man-pages/man2/mmap.2.html

I think a comment in the header file, with example, would be really useful.